### PR TITLE
SAI, part 2

### PR DESCRIPTION
**Pull request by gdamron**
Updated Visuals:
- Import and use assets from original demo
- Refine text display, including placement, drop shadow, and font
- Add fade transition for text

Animation:
- Incorporate existing animations into basic state
- Proactively attract players' attention when highlighting an element.

Script API:
- Add several schema properties for customizing attraction, motion, and text display 
- Add `animate` for triggering animation states
- Support optional timeout for text

### CODE REVIEW

File: Assets/Source/Player/IUX/Widget/SaiWidget.cs [hunk 1]
```diff
@@ -164,49 +222,95 @@ protected override void LoadInternalAfterChildren()
         {
             base.LoadInternalAfterChildren();
 
+            // Text display
+            _textPrimitive = _primitives.Text(Schema);
+
+            _textPrimitive.HorizontalOverflow = HorizontalWrapMode.Wrap;
+            _textPrimitive.VerticalOverflow = VerticalWrapMode.Overflow;
+            _textPrimitive.Alignment = TextAlignmentType.BotCenter;
```
Comment (thegoldenmule) [inline Assets/Source/Player/IUX/Widget/SaiWidget.cs]: Aren't these in schema?

File: Assets/Source/Player/Scripting/Interface/App/SaiJsApi.cs [hunk 2]
```diff
@@ -65,22 +65,85 @@ public void speak(string text, IJsCallback callback)
                 return;
             }
 
-            _sai.Speak(text, () => callback.Invoke());
+            _sai.Speak(text, () => callback?.Invoke());
```
Comment (thegoldenmule) [inline Assets/Source/Player/Scripting/Interface/App/SaiJsApi.cs]: Anytime we call a JS callback, we should wrap it in a try/catch so it doesn't kill the C# side.

File: Assets/Source/Player/Scripting/Interface/App/SaiJsApi.cs [hunk 3]
```diff
@@ -65,22 +65,85 @@ public void speak(string text, IJsCallback callback)
...
+        /// <param name="text">The text to display</param>
+        /// <param name="callback">Called after text disappears</param>
+        public void write(string text, IJsCallback callback)
+        {
+            if (!enabled)
+            {
+                return;
+            }
+
+            _sai.Write(text, float.MaxValue, () => callback?.Invoke());
```
Comment (thegoldenmule) [inline Assets/Source/Player/Scripting/Interface/App/SaiJsApi.cs]: Same here

File: Assets/Source/Player/Scripting/Interface/App/SaiJsApi.cs [hunk 4]
```diff
@@ -65,22 +65,85 @@ public void speak(string text, IJsCallback callback)
...
+        /// <param name="callback">Called after text disappears</param>
+        public void write(string text, int timeoutMs, IJsCallback callback)
+        {
+            if (!enabled)
+            {
+                return;
+            }
+
+            var seconds = timeoutMs / (float) TimeUtils.SECOND_IN_MS;
+            _sai.Write(text, seconds, () => callback?.Invoke());
```
Comment (thegoldenmule) [inline Assets/Source/Player/Scripting/Interface/App/SaiJsApi.cs]: And here

File: Assets/Source/Player/IUX/Widget/SaiWidget.cs [hunk 5]
```diff
@@ -147,6 +176,35 @@ public void Speak(string text, Action callback = null)
             _renderer.Speak(text, callback);
         }
 
+        /// <summary>
+        /// Display text.
+        /// </summary>
+        /// <param name="text"></param>
+        /// <param name="timeout"></param>
+        /// <param name="callback"></param>
+        public void Write(string text, float timeout = float.MaxValue, Action callback = null)
```
[the remark this answers is not in the document]
Comment (gdamron) [reply Assets/Source/Player/IUX/Widget/SaiWidget.cs]: 👍

File: Assets/Source/Player/IUX/Widget/SaiWidget.cs (hunk 1 above)
Comment (gdamron) [reply Assets/Source/Player/IUX/Widget/SaiWidget.cs]: 👍

File: Assets/Source/Player/Scripting/Interface/App/SaiJsApi.cs (hunk 2 above)
Comment (gdamron) [reply Assets/Source/Player/Scripting/Interface/App/SaiJsApi.cs]: 👍

File: Assets/Source/Player/Scripting/Interface/App/SaiJsApi.cs (hunk 3 above)
Comment (gdamron) [reply Assets/Source/Player/Scripting/Interface/App/SaiJsApi.cs]: 👍

File: Assets/Source/Player/Scripting/Interface/App/SaiJsApi.cs (hunk 4 above)
Comment (gdamron) [reply Assets/Source/Player/Scripting/Interface/App/SaiJsApi.cs]: 👍

**Comment by gdamron** [comment]
Yes, @jsundra, I'll add JS schema access. should be an easy change.

**Comment by thegoldenmule** [comment]
@gdamron Please submit a corresponding PR to the js docs!

**Comment by gdamron** [comment]
PR for docs
https://github.com/enklu/enkluplayer-jsapi-docs/pull/25

**Pull request merged**